[fx] give warning instead of fatal the program when submod not found during adding get_attr

842974287 opened the pull request:
Summary:
Currently when create get_attr node, if the attribute is in a submodule, we'll fist find the submodule. If the submodule isn't in the owning module we throw an exception.

However, if the attribute can't be found, we give a warning but still allow to create the get_attr node. To align with this behavior, we change the reaction when submod not found from fatal to giving a warning.

Test Plan: CI

Differential Revision: D31021535

jamesr66a commented on torch/fx/graph.py:
```diff
@@ -561,7 +561,11 @@ def get_attr(self, qualified_name: str, type_expr: Optional[Any] = None) -> Node
         def _get_attr_reference_exists(mod: torch.nn.Module, qualified_name: str) -> bool:
```
This whole function looks like it was written with the assumption that `Module.get_submodule` returns None if that submodule does not exist, but that is not the case if you look at the function:

https://github.com/pytorch/pytorch/blob/992dad18552883224016d53429c48f2e932651cf/torch/nn/modules/module.py#L390

Alternative fixes would be:
1) Make get_submodule conform to that behavior
2) Add a `Module.find_submodule` method that `get_submodule` would be implemented in terms of. Then, `get_submodule` is just a wrapper that additionally throws if it gets a None value from `find_submodule`

I'm not sure if we want to do more intrusive changes to `nn.Module` so I'm going to just approve this PR for now

jamesr66a replied:
But given the above, can you update `submod` to be typed ad `nn.Module` and delete the code below that handles if `submod` is None?

842974287 committed:
[fx] give warning instead of fatal the program when submod not found during adding get_attr (#65225)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/65225

Currently when create get_attr node, if the attribute is in a submodule, we'll fist find the submodule. If the submodule isn't in the owning module we throw an exception.

However, if the attribute can't be found, we give a warning but still allow to create the get_attr node. To align with this behavior, we change the reaction when submod not found from fatal to giving a warning.

Test Plan: CI

Reviewed By: jamesr66a, jfix71

Differential Revision: D31021535

fbshipit-source-id: 0d872de4872bbd65e8be52952648dd591096e033
